ext/storage_sources/local_store/local_store.c

```diff
@@ -0,0 +1,1220 @@
+
+    ret = stat(path, &sb);
+    if (ret == 0)
+        *sizep = sb.st_size;
+    else
+        ret = local_err(local, session, errno, "%s: ss_size stat", path);
+
+err:
+    free(path);
+    return (ret);
+}
+
+/*
+ * local_terminate --
+ *     Discard any resources on termination
+ */
+static int
+local_terminate(WT_STORAGE_SOURCE *storage, WT_SESSION *session)
+{
+    LOCAL_FILE_HANDLE *local_fh, *safe_fh;
+    LOCAL_STORAGE *local;
+    int ret;
+
+    ret = 0;
+    local = (LOCAL_STORAGE *)storage;
+
+    local->op_count++;
+
+    /*
+     * We should be single threaded at this point, so it is safe to destroy the lock and access the
+     * file handle list without it.
+     */
+    if ((ret = pthread_rwlock_destroy(&local->file_handle_lock)) != 0)
+        (void)local_err(local, session, ret, "terminate: pthread_rwlock_destroy");
+
+    TAILQ_FOREACH_SAFE(local_fh, &local->fileq, q, safe_fh)
+    local_file_close_internal(local, session, local_fh, false);
+
+    free(local);
+    return (ret);
+}
+
+/*
+ * local_file_close --
+ *     ANSI C close.
+ */
+static int
+local_file_close(WT_FILE_HANDLE *file_handle, WT_SESSION *session)
+{
+    LOCAL_STORAGE *local;
+    LOCAL_FILE_HANDLE *local_fh;
+    LOCAL_FLUSH_ITEM *flush;
+    int ret, t_ret;
+
+    ret = 0;
+    local_fh = (LOCAL_FILE_HANDLE *)file_handle;
+    local = local_fh->local;
+
+    local->fh_ops++;
+    if ((ret = pthread_rwlock_wrlock(&local->file_handle_lock)) != 0)
+        /* There really isn't anything more we can do. It will get cleaned up on terminate. */
+        return (local_err(local, session, ret, "file handle close: pthread_rwlock_wrlock"));
+
+    TAILQ_REMOVE(&local->fileq, local_fh, q);
+
+    if ((ret = pthread_rwlock_unlock(&local->file_handle_lock)) != 0)
+        (void)local_err(local, session, ret, "file handle close: pthread_rwlock_unlock");
+
+    /*
+     * If we need to track flushes for this file, save the flush item on our queue.
+     */
+    if (ret == 0 && ((flush = local_fh->flush)) != NULL) {
+        if ((ret = pthread_rwlock_wrlock(&local->flush_lock)) != 0)
+            (void)local_err(local, session, ret, "file handle close: pthread_rwlock_wrlock2");
+
+        if (ret == 0) {
+            TAILQ_INSERT_HEAD(&local->flushq, flush, q);
+            local_fh->flush = NULL;
+
+            if ((ret = pthread_rwlock_unlock(&local->flush_lock)) != 0)
+                (void)local_err(local, session, ret, "file handle close: pthread_rwlock_unlock2");
+            if (ret == 0 && ((flush->src_path = strdup(local_fh->path)) == NULL))
+                ret = ENOMEM;
+        }
+    }
+
+    if ((t_ret = local_file_close_internal(local, session, local_fh, true)) != 0) {
+        if (ret == 0)
+            ret = t_ret;
+    }
+
+    return (ret);
+}
+
+/*
+ * local_file_close_internal --
+ *     Internal file handle close.
+ */
+static int
+local_file_close_internal(
+  LOCAL_STORAGE *local, WT_SESSION *session, LOCAL_FILE_HANDLE *local_fh, bool normal)
```

Comment:
This is super minor so ignore if you want. In most WT functions that have this sort of boolean, it is flipped to be `final`.